3 lab(work_with_object_collection): реализовал лабораторную 3 по предмету ООАиП.

RockInHead opened the pull request.

RockInHead committed:
feature(MainWindow.xaml): сделал макет верстки для 3ей лабораторной.

RockInHead committed:
feature(View): реализована логика кнопок. Требуется рефакторинг!

- реализованы кнопки: добавления, удаления, редактирования и подтверждения для работы с списком контактов.
- реализовано отображение списка контактов в листбокс.

RockInHead committed:
feature(View): реализовал сохранение данных в JSON. Требуется рефакторинг!

RockInHead committed:
feature(View): реализовал логику не сохранения изменений при выборе другого элемента из списка.

RockInHead committed:
refactoring(View): Сделал рефакторинг всего проекта. Удалил лишнее и добавил xml комментарии.

RockInHead committed:
feature(MainWindow.xaml): Сделал ограничение по размерам для окна.

sandra155464 commented on Programming/Contacts/Contacts/MainWindow.xaml:
```diff
@@ -2,124 +2,176 @@
-                    HorizontalAlignment="Stretch"
-                    Text="{Binding PhoneNumber, UpdateSourceTrigger=PropertyChanged}"/>
+                    <!--  Поле поиска  -->
+                    <Grid Grid.Row="0" Margin="3 3 5 3">
```
Сделать Margin = 3

sandra155464 commented on Programming/Contacts/Contacts/MainWindow.xaml:
```diff
@@ -2,124 +2,176 @@
+                    <ListBox
+                        x:Name="ContactsList"
+                        Grid.Row="1"
+                        Margin="3 7 3 3"
```
Также сделать Margin = 3

sandra155464 commented on Programming/Contacts/Contacts/MainWindow.xaml:
```diff
@@ -2,124 +2,176 @@
+                        ItemsSource="{Binding Contacts}"
+                        SelectedItem="{Binding SelectedContact, Mode=TwoWay}"/>
+                    <!--  Кнопки управления  -->
+                    <Grid Grid.Row="2" Margin="3 3 3 3">
```
Если отступ одинаковый со всех сторон, то можно написать цифру "3" лишь один раз - в прошлой лабораторной было аналогичное замечание

sandra155464 commented on Programming/Contacts/Contacts/MainWindow.xaml.cs:
```diff
@@ -1,4 +1,5 @@
-﻿using System.Windows;
+﻿using System.Collections.ObjectModel;
```
Убрать неиспользуемый using, лишние пустые строки, а также закомментированную привязку данных

sandra155464 commented on Programming/Contacts/Contacts/MainWindow.xaml:
```diff
@@ -2,124 +2,176 @@
-            </Grid>
-        </GroupBox>
+                    <!--  Кнопка "Apply"  -->
+                    <Button
```
В методичке было написано, что кнопка Apply при добавлении/редактировании должна **появляться**, а при сохранении результатов - **исчезать**. Нужно привязать свойство Visible кнопки к булевому свойству из MainVM с использованием конвертера величин (про это в методичке тоже было). У тебя в MainVM есть свойство IsApplyButtonVisible, но привязывать нужно не к нему, а к свойству IsAddOrEditMode - чуть ниже напишу, откуда оно появилось

sandra155464 commented on Programming/Contacts/Contacts/Model/Services/ContactSerializer.cs:
```diff
@@ -2,60 +2,52 @@
+    /// Статический конструктор для инициализации пути к файлу.
     /// </summary>
-    public static void CreateDirectory()
+    static ContactSerializer()
```
У конструктора должен быть прописан модификатор доступа public

sandra155464 replied:
У конструктора должен быть модификатор доступа public.
И почему используется конструктор? В прошлой лабе ты использовал метод по созданию пути сохранения файла

sandra155464 commented on Programming/Contacts/Contacts/Model/Services/ContactSerializer.cs:
```diff
@@ -2,60 +2,52 @@
+    /// Возвращает список контактов, если файл существует и успешно десериализован.
+    /// В противном случае возвращает пустой список.
+    /// </returns>
+    public static List<ContactVM> LoadContacts()
```
Согласно паттерну MVVM, сущности уровня Model не должны ничего знать о сущностях уровня ViewModel. Здесь же у тебя модельный класс возвращает список объектов ContactVM, и такого быть не должно. Чтобы это исправить, можно либо как-то преобразовывать коллекцию типа ContactVM к коллекции типа Contact в MainVM при передаче в сериалайзер, либо реализовать интерфейс INotifyPropertyChanged непосредственно в классе Contact, и в MainVM использовать его

sandra155464 replied:
При этом во втором случае класс Contact останется на уровне Model, даже с реализованным INotifyPropertyChanged, так что нарушения связи между уровнями не будет

sandra155464 commented on Programming/Contacts/Contacts/ViewModel/ContactVM.cs:
```diff
@@ -0,0 +1,98 @@
+    }
+
+    /// <summary>
+    /// Событие, уведомляющее об изменениях в свойствах.
```
Это событие появилось в классе из-за реализации интерфейса, поэтому вместо того, чтобы писать к нему XML-комментарий вручную, можно просто написать:
`/// <inheritdoc cref="INotifyPropertyChanged.PropertyChanged"/>`

sandra155464 commented on Programming/Contacts/Contacts/ViewModel/MainVM.cs:
```diff
@@ -10,22 +13,134 @@ public class MainVM : INotifyPropertyChanged
     /// </summary>
     private Contact _currentContact;
```
Поскольку теперь у тебя есть SelectedContact, то это поле и соответствующее ему свойство не нужны, их нужно убрать

sandra155464 replied:
Также нужно убрать свойства Name, Email и PhoneNumber, они тоже нигде не используются

sandra155464 commented on Programming/Contacts/Contacts/ViewModel/MainVM.cs:
```diff
@@ -10,22 +13,134 @@ public class MainVM : INotifyPropertyChanged
     {
         _currentContact = new Contact();
+        Contacts = new ObservableCollection<ContactVM>(ContactSerializer.LoadContacts());
 
```
Лишняя пустая строка

sandra155464 commented on Programming/Contacts/Contacts/ViewModel/RelayCommand.cs:
```diff
@@ -0,0 +1,57 @@
+    /// </summary>
+    public event EventHandler CanExecuteChanged
+    {
+        add { CommandManager.RequerySuggested += value; }
```
Согласно внутреннему стандарту оформления кода, фигурные скобки пишутся каждая с новой строки, и каждая строка внутри них тоже

sandra155464 commented on Programming/Contacts/Contacts/ViewModel/RelayCommand.cs:
```diff
@@ -0,0 +1,57 @@
+﻿using System.Windows.Input;
+
+/// <summary>
+/// Реализует интерфейс <see cref="ICommand"/> для привязки команд в WPF, позволяя определить действия для выполнения 
```
Эта строка в XML-комментарии класса превышает по длине 100 символов, нужно перенести часть слов на новую строку

sandra155464 commented on Programming/Contacts/Contacts/ViewModel/MainVM.cs:
```diff
@@ -10,22 +13,134 @@ public class MainVM : INotifyPropertyChanged
     /// </summary>
     public MainVM()
     {
         _currentContact = new Contact();
```
В классе Contact при вызове конструктора без параметров (`Contact()`) в свойства контакта присваиваются данные из прошлой лабораторной, и это присваивание нужно убрать, оставив конструктор пустым

sandra155464 replied:
Замечание немножко не исправлено: конструктор без параметров можно оставить вообще пустым. Вот так:
`
public Contact()
{

}
`

sandra155464 replied:
Замечание немножко не исправлено: конструктор без параметров можно оставить вообще пустым. Вот так:
``` cs
public Contact()
{

}
```

Или что-то в программе ломается, если поля контакта null, а не пустые строки? Я локально попробовала изменить, и у меня всё работает нормально )

sandra155464 commented on Programming/Contacts/Contacts/ViewModel/MainVM.cs:
```diff
@@ -10,22 +13,134 @@ public class MainVM : INotifyPropertyChanged
+    /// Команда для добавления нового контакта.
+    /// </summary>
+    public ICommand AddCommand { get; }
+
```
Здесь и ниже лишние пустые строки. Свойства отделяются друг от друга одной пустой строкой

sandra155464 commented on Programming/Contacts/Contacts/ViewModel/MainVM.cs:
```diff
@@ -101,36 +216,130 @@ public string Email
-}
 
+    /// <summary>
+    /// Вызывает событие <see cref="PropertyChanged"/> для обновления интерфейса.
```
Здесь тоже можно использовать `<inheritdoc/>`

sandra155464 replied:
Прошу прощения, недоглядела немного в прошлый раз. У метода OnPropertyChanged() в XML-комментарии не нужно никакого `<inheritdoc/>`, т.к. это не наследуемый при реализации член интерфейса. `<inheritdoc/>` надо сделать у события PropertyChanged, как в классе Contact

sandra155464 commented on Programming/Contacts/Contacts/ViewModel/MainVM.cs:
```diff
@@ -101,36 +216,130 @@ public string Email
+        {
+            SelectedContact = null;
+        }
+        ContactSerializer.SaveContacts(Contacts);
```
После фигурной скобочки нужна одна пустая строка

sandra155464 commented on Programming/Contacts/Contacts/ViewModel/MainVM.cs:
```diff
@@ -101,36 +216,130 @@ public string Email
+    public void AddContact(object parameter)
     {
-        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
+        SelectedContact = null;
```
Не совсем понимаю, зачем нужна эта строка - вроде как без неё добавление тоже корректно работает

RockInHead replied:
Решил все-таки оставить эту строку кода.

Вот описание, что будет если убрать эту строчку кода:
Если пользователь выберет контакт и после нажмет кнопку Add, то в ListBox контакт останется по умолчанию выбранным, хотя TextBoxы очистятся для нового контакта. Так вот если при заполнении выбрать этот изначальный контакт, то поля не отобразят контакт из списка, а будет продолжен режим создания нового контакта, а это противоречит тз.

https://github.com/user-attachments/assets/3b9173e1-a89d-4a9d-b861-df46ea91584b

Если не совсем понятно объяснил, пиши)

sandra155464 replied:
Хорошо, теперь поняла ;)

sandra155464 commented on Programming/Contacts/Contacts/ViewModel/MainVM.cs:
```diff
@@ -101,36 +216,130 @@ public string Email
+    /// </summary>
+    /// <param name="parameter">Параметр команды.</param>
+    /// <returns>Возвращает <c>true</c>, если контакт можно добавить; иначе <c>false</c>.</returns>
+    private bool CanAddContact(object parameter) => !IsApplyButtonVisible;
```
Ещё по этому классу у меня много замечаний по полям и свойствам, но их лучше выписать все вместе, чтобы было понятнее:
1) убрать поле _isEditingContact, оставить только автосвойство
2) автосвойство IsEditingContact переименовать на IsAddOrEditMode. У тебя уже есть свойство IsReadOnlyMode, и слово "Mode" в его названии хорошо отражает, когда главное окно находится в состоянии только чтения. По аналогии лучше назвать "IsAddOrEditMode" свойство, указывающее на то, что главное окно находится в состоянии добавления или редактирования контакта. Объединение состояний добавления контакта и редактирования существующего под одним свойством также не будет ошибкой, поскольку для вёрстки это одно и то же состояние - кнопки Add, Edit и Remove блокируются, поля справа становятся доступными для ввода, появляется кнопка Apply
3) в текущей реализации значения предикатов CanAddContact, CanEditContact, CanRemoveContact и CanApplyContact частично или полностью зависят от значения свойства IsApplyButtonVisible. Если вдуматься, то получается, что возможность добавления контакта, его изменения, удаления и применения изменений зависит от видимости кнопочки - не очень прослеживается связь. При вычислении значений этих предикатов свойство IsApplyButtonVisible можно спокойно заменить на свойство IsAddOrEditMode - это будет так же корректно работать, но ещё и будет чуть более логично
4) если свойство IsApplyButtonVisible можно вообще не использовать, и всё будет работать так же хорошо, значит, его можно также убрать :)
5) свойства IsReadOnlyMode и IsAddOrEditMode противоположны друг другу по значению: если IsReadOnlyMode=true, то IsAddOrEditMode=false, и наоборот. Можно сделать так, чтобы значение одного свойства вычислялось из значения другого. Я предлагаю сделать значение IsAddOrEditMode зависимым от IsReadOnlyMode, поскольку IsReadOnlyMode - это такое "дефолтное" состояние главного окна, при первом запуске приложения оно находится именно в нём. Соответственно, нужно убрать поле _isAddOrEditMode, а в соответствующем свойстве возвращать (сеттера не будет) инверсию от IsReadOnlyMode. Отдельно стоит подумать, как сделать так, чтобы при изменении значения IsReadOnlyMode также обновлялось и значение IsAddOrEditMode

sandra155464 commented:
Оставила замечания по лабе. Я не помню, чтобы видела ошибки где-то в классах, но перед отправкой коммитов с исправлениями ещё проверь, на всякий случай, порядок расположения членов класса

RockInHead committed:
style(MainWindow.xaml, MainWindow.xaml.cs): поправил Margin, удалил неиспользуемый using и лишние комментарии.

RockInHead committed:
style(View): исправил ошибки с комментариями и оформлением во всём коде, изменил логику конструктора по умолчанию для Contact.cs.

RockInHead committed:
refactoring(View): удалил класс ContactVM и реализовал интерфейс INotifyPropertyChanged непосредственно в классе Contact. Также удалил все неиспользуемые поля и свойства с прошлой лабы.

RockInHead committed:
fix(Contact): исправил баг, при изменении контакта наименование элемента в ListBox не менялось. Поправил, теперь работает корректно.

RockInHead committed:
refactoring(ContactSerializer, MainVM): по аналогии с прошлой лабой реализовал функцию для создания каталога, где будут сохраняться контакты.

RockInHead committed:
refactoring(MainVM): удалил поле _isEditingContact, оставил только автосвойство.

RockInHead committed:
style(MainVM): переименовал автосвойство IsEditingContact в IsAddOrEditMode.

RockInHead committed:
refactoring(MainVM): заменил свойство IsApplyButtonVisible на свойство IsAddOrEditMode при вычислении предикатов.

RockInHead committed:
refactoring(MainVM): удалил всё, что касалось свойства IsApplyButtonVisible, без него работает вполне корректно.

RockInHead committed:
feature(View): создал конвертер значений для отображения и скрытия кнопки Apply.

RockInHead committed:
refactoring(MainVM): убрал лишний комментарий и поправил функционал на кнопку Add.

- подробнее о кнопки Add написал в комментарии на GitHub на замечание по этой строчке кода.

sandra155464 commented on Programming/Contacts/Contacts/Model/Contact.cs:
```diff
@@ -1,17 +1,34 @@
+    private string _phoneNumber;
+
     /// <summary>
     /// Конструктор класса Contact. 
```
Строка XML-комментария не должна заканчиваться пробелами

sandra155464 commented on Programming/Contacts/Contacts/MainWindow.xaml.cs:
```diff
@@ -17,9 +16,6 @@ public MainWindow()
 
```
Лишняя пустая строка

sandra155464 commented on Programming/Contacts/Contacts/ViewModel/RelayCommand.cs:
```diff
@@ -1,7 +1,8 @@
 /// <summary>
-/// Реализует интерфейс <see cref="ICommand"/> для привязки команд в WPF, позволяя определить действия для выполнения 
+/// Реализует интерфейс <see cref="ICommand"/> для привязки команд в WPF,
+/// позволяя определить действия для выполнения 
```
Снова лишний пробел в конце строки XML-комментария

sandra155464 commented on Programming/Contacts/Contacts/Model/Services/ContactSerializer.cs:
```diff
@@ -9,25 +9,35 @@ public static class ContactSerializer
+        if (contacts == null)
+        {
+            throw new ArgumentNullException(nameof(contacts), "Контакт не может быть null.");
+        }
```
Если после закрывающей фигурной скобки есть ещё код, то между ним и закрывающей фигурной скобкой должна быть одна пустая строка

sandra155464 replied:
Замечание не исправлено: перед вызовов метода CreateDirectory() должна быть одна пустая строка, т.к. вызов метода идёт после закрывающейся фигурной скобки

sandra155464 commented on Programming/Contacts/Contacts/ViewModel/MainVM.cs:
```diff
@@ -78,19 +59,17 @@ public MainVM()
-            _isEditingContact = value;
-            OnPropertyChanged(nameof(IsEditingContact));
-        }
+            return !IsReadOnlyMode;
```
Можно переписать чуть лаконичнее:
``` cs
public bool IsAddOrEditMode => !IsReadOnlyMode;
```

И ещё стоит поправить XML-комментарий у свойства. Раз оно теперь доступно только на get, то комментарий должен содержать только слово "возвращает"

sandra155464 commented on Programming/Contacts/Contacts/ViewModel/MainVM.cs:
```diff
@@ -258,8 +149,7 @@ public void RemoveContact(object parameter)
+        SelectedContact = new Contact();
         IsReadOnlyMode = false;
     }
 
```
В методе ниже после строки bindingGroup.CommitEdit(); лишняя пустая строка. Конструкции if, for, while и т.д. не отделяются пустыми строками "спереди", только после фигурной скобочки, и то если после них написан ещё какой-то код

sandra155464 replied:
Здесь замечание тоже не исправлено: после bindingGroup.CommitEdit(); пустая строка не нужна, а вот перед ним - нужна. Может, я объяснила непонятно, поэтому, на всякий случай, укажу, как это должно выглядеть в коде:
``` cs
if (parameter is not BindingGroup bindingGroup)
{
    return;
}

bindingGroup.CommitEdit();
if (SelectedContact == null)
{
    return;
}
```

sandra155464 commented on Programming/Contacts/Contacts/ViewModel/MainVM.cs:
```diff
@@ -286,16 +176,11 @@ public void ApplyContact(object parameter)
-    /// Вызывает событие <see cref="PropertyChanged"/> для обновления интерфейса.
-    /// </summary>
-    /// <param name="propertyName">Имя измененного свойства.</param>
+    /// <inheritdoc cref="INotifyPropertyChanged.PropertyChanged"/>
```
Это XML-комментарий для события, а само событие объявлено на 37 строке. К событию надо написать <inheritdoc/>, а тут написать комментарий вручную

sandra155464 commented:
Закрыла часть старых замечаний, оставила несколько новых

RockInHead committed:
refactoring(View): закрыл новые и часть старых замечаний.

sandra155464 commented:
Закрыла вчерашние замечания, осталось поправить всего пару моментов на оформление кода. Как исправишь - можно закрывать обсуждения и мёрджить пулл-реквест

RockInHead committed:
style(View): исправил последние правки по коду.
